src/hotspot/share/oops/oopCast.inline.hpp

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef SHARE_OOPS_OOPCAST_INLINE_HPP
+#define SHARE_OOPS_OOPCAST_INLINE_HPP
+
+#include "oops/oopsHierarchy.hpp"
+#include "oops/oop.inline.hpp"
+
+template<typename T>
+static bool is_oop_type(oop theOop) {
+  static_assert(sizeof(T) == 0, "No is_oop_type specialization found for this type");
```

Comment:
The `sizeof(T) == 0` check always evaluates to false. But, if we were to use just `false` instead we would always get a compilation error here. Having the check depend on `T` means it is only run when the template is instantiated (with a type that is not supported by one of the specializations below).